examples/scripts/denoise_cplx_tv_nlpadmm.py

```diff
@@ -0,0 +1,246 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# This file is part of the SCICO package. Details of the copyright
+# and user license can be found in the 'LICENSE.txt' file distributed
+# with the package.
+
+r"""
+Complex Total Variation Denoising with NLPADMM Solver
+=====================================================
+
+This example demonstrates solution of a problem of the form
+
+$$\argmin_{\mathbf{x}} \; f(\mathbf{x}) + g(\mathbf{z}) \; \text{such
+ that}\; H(\mb{x}, \mb{z}) = 0 \;,$$
+
+where $H$ is a function, via a variant of the proximal ADMM
+algorithm for problems with a non-linear operator constraint
+:cite:`benning-2016-preconditioned`. The example problem represents
+total variation (TV) denoising applied to a complex image with
+piece-wise smooth magnitude and non-smooth phase. The appropriate TV
+denoising formulation for this problem is
+
+$$\mathrm{argmin}_{\mathbf{x}} \; (1/2) \| \mathbf{y} - \mathbf{x}
+\|_2^2 + \lambda \| C(\mathbf{x}) \|_{2,1} \;,$$
+
+where $\mathbf{y}$ is the measurement, $\|\cdot\|_{2,1}$ is the
+$\ell_{2,1}$ mixed norm, and $C$ is a non-linear operator that applies
+a linear difference operator to the magnitude of a complex array.
+This problem is represented in the form above by taking $H(\mb{x},
+\mb{z}) = C(\mathbf{x}) - \mb{z}$. The standard TV solution, which is
+also computed for comparison purposes, gives very poor results since
+the difference is applied independently to real and imaginary
+components of the complex image.
+"""
+
+
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+from xdesign import SiemensStar, discrete_phantom
+
+import scico.numpy as snp
+import scico.random
+from scico import function, functional, linop, loss, metric, operator, plot
+from scico.examples import phase_diff
+from scico.optimize import NonLinearPADMM, ProximalADMM
+from scico.util import device_info
+
+"""
+Create a ground truth image.
+"""
+N = 256  # image size
+phantom = SiemensStar(16)
+x_mag = snp.pad(discrete_phantom(phantom, N - 16), 8) + 1.0
+x_mag /= x_mag.max()
+# Create reference image with structured magnitude and random phase
+x_gt = x_mag * snp.exp(-1j * scico.random.randn(x_mag.shape, seed=0)[0])
+
+
+"""
+Add noise to create a noisy test image.
+"""
+σ = 0.25  # noise standard deviation
+noise, key = scico.random.randn(x_gt.shape, seed=1, dtype=snp.complex64)
+y = x_gt + σ * noise
+
+
+"""
+Denoise with standard total variation.
+"""
+λ_tv = 6e-2
+f = loss.SquaredL2Loss(y=y)
+g = λ_tv * functional.L21Norm()
+# The append=0 option makes the results of horizontal and vertical finite
+# differences the same shape, which is required for the L21Norm.
+C = linop.FiniteDifference(input_shape=y.shape, input_dtype=snp.complex64, append=0)
+# Constraint function imposing z = C(x) constraint
+H = function.Function(
```

Comment:
H does not seem to be used and then re-defined later below.

Reply:
Good catch. Fixed.